src/ops/reshape.cc

```diff
@@ -34,7 +35,20 @@ using Legion::Task;
 using Legion::TaskArgument;
 using Legion::TaskLauncher;
 
-ReshapeParams::ReshapeParams(std::vector<int> const &_shape) : shape(_shape) {}
+/* Params */
+bool operator==(ReshapeParams const &lhs, ReshapeParams const &rhs) {
+  return lhs.shape == rhs.shape;
+}
+
+bool ReshapeParams::is_valid(ParallelTensorShape const &input) const {
+  return input.is_valid();
```

Comment:
There should probably also at least be a volume check if not some others here against the shape field in `ReshapeParams`

Reply:
Maybe I am going about the validity checks incorrectly, so this would be helpful to clarify. I've been referring to the [original function](https://github.com/flexflow/FlexFlow/blob/master/src/ops/reshape.cc#L426) for the right checks to make. It seems like there are some `assert` calls in the constructor on the actual `ParallelTensor` input volume. If we are still doing those, then do we need additional volume checks on the shape in node creation?

Reply:
`[...]Params::is_valid` is used to determine the return value of `Model::get_or_create_node<T>`, which is then used in `GraphXfer::create_new_operator`, which is called from `Graph::find_matches` to determine if a substitution applys to an input graph. Therefore, if a check is missing from `is_valid`, instead of rejecting that substitution and going and explore other ones, we'll apply that substitution, construct the operator, fail the assertion, and crash out of search (which is bad). Thus the checks in `is_valid` should generally be a superset of the assertions in the operator itself except in cases where a check would indicate that something is more wrong than just an invalid subsetitution being tried out. 

This case probably won't happen too frequently as the substitution generator is in theory supposed to only produce substitutions where the src ops and constraints matching implies that the dst ops and constraints should always work, so this is why I'm guessing this doesn't come up as often as it otherwise would. That said, I think it's probably a good behavior to maintain from now and as the generator -> Unity pipeline gets refined more we can consider potentially changing it. @jiazhihao Correct me if I'm forgetting anything here.

 > I've been referring to the [original function](https://github.com/flexflow/FlexFlow/blob/master/src/ops/reshape.cc#L426) for the right checks to make
 
 Unfortunately the original functions frequently only include a subset of the valid checks.

Reply:
That makes sense. So if the original `get_or_create_[op]_node` functions only contained a subset of the valid checks, then we probably need to look at `is_valid` for each operator more closely and see if there are more checks required.

Reply:
Just to write down what we discussed in the meeting: except for the graph cycle checks the substitution generator should be responsible for maintaining correctness through substitution application, not the runtime. As such we should eventually remove the `is_valid` checks (#301)